crates/hotshot/src/traits/networking/combined_network.rs

```diff
@@ -165,20 +168,63 @@ impl<TYPES: NodeType> CombinedNetworks<TYPES> {
 
         if !primary_failed && Self::should_delay(&message) {
             let duration = *self.delay_duration.read().await;
+            let primary_failed = Arc::clone(&self.primary_down);
             let primary_fail_counter = Arc::clone(&self.primary_fail_counter);
-            self.delayed_tasks
+            let mut receiver = self
+                .delayed_tasks_channels
                 .write()
                 .await
                 .entry(message.kind.view_number().u64())
-                .or_default()
-                .push(async_spawn(async move {
-                    async_sleep(duration).await;
-                    info!("Sending on secondary after delay, message possibly has not reached recipient on primary");
-                    primary_fail_counter.fetch_add(1, Ordering::Relaxed);
-                    secondary_future.await
-                }));
+                .or_insert_with(|| {
+                    let (s, r) = broadcast(1);
+                    (s, r.deactivate())
+                })
+                .1
+                .activate_cloned();
+            async_spawn(async move {
+                async_sleep(duration).await;
+                if receiver.try_recv().is_ok() {
+                    debug!(
+                        "Not sending on secondary after delay, task was canceled in view update"
+                    );
+                    match primary_fail_counter.load(Ordering::Relaxed) {
```

Comment:
I think I'm finding this logic a little confusing, Maybe it's the nested ifs.  My reading of it is when the primary is up we keep track of failures based on if the view changes in time or not.  Then the next time we try to send we check this number and set the primary as down if it's too high.  Then we proceed to send to the secondary with no delay and after some number of messages we try the primary again.  If it succeeds it will be 1 below the threshold to set the primary down, but when do we set `primary_failed` back to false?  Seems like it's only after trying with a delay until the counter is reduced to 0.  But I don't see how that can happen because with `primary_failed` true we don't hit the code that reduces `primary_fail_counter`.

I'm probably missing something but I think we should try to come up with a way to make this logic more straightforward to read

Reply:
Ok I think I see `primary_down` and `primary_failed` are the same thing.  I think I convinced myself the logic is correct but it's still a bit confusing to me

Reply:
I think just a few comments for now at each if block is completely fine.  I am convinced the logic is correct based on your testing and reading the code